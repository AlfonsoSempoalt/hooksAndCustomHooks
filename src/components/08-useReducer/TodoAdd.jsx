import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from '../../hooks/useForm';

const TodoAdd = ({ handleAddTodo }) => {
  const [formValues, handleInputChange, resetForm] = useForm({
    description: '',
  });
  const { description } = formValues;

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length >= 2) {
      const newTodo = {
        id: new Date().getTime(),
        desc: description,
        done: false,
      };
      handleAddTodo(newTodo);
      resetForm();
    }
  };

  return (
    <>
      <h4>Add todo</h4>
      <hr />
      <form onSubmit={handleSubmit} className='d-grid gap-2'>
        <input
          type='text'
          placeholder='learn...'
          name='description'
          value={description}
          onChange={handleInputChange}
          autoComplete='off'
          className='form-control'
        />
        <button
          type='submit'
          className='btn btn-outline-primary mt-2 btn-block'
        >
          Add
        </button>
      </form>
    </>
  );
};

TodoAdd.propTypes = {
  handleAddTodo: PropTypes.func.isRequired,
};

export default TodoAdd;
