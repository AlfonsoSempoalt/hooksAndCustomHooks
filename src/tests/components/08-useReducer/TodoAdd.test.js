import React from 'react';
import { shallow } from 'enzyme';
import TodoAdd from '../../../components/08-useReducer/TodoAdd';

describe('tests to todoAdd', () => {
    const handleAddTodo = jest.fn();
    const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

    test('should match with the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('shouldnt call handleAddTodo', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault() {} });
        expect(handleAddTodo).not.toHaveBeenCalled();
    });

    test('should have been call once handleAddTodo', () => {
        const value = 'leer ruso';
        wrapper.find('.form-control').simulate('change', {
            target: {
                value: value,
                name: 'description',
            },
        });
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault() {} });
        expect(handleAddTodo).toHaveBeenCalled();
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false,
        });
    });

    test('should have been call reset func', () => {
        const value = 'leer ruso';
        wrapper.find('.form-control').simulate('change', {
            target: {
                value: value,
                name: 'description',
            },
        });
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault() {} });
        expect( wrapper.find('input').prop('value')).toBe('');
    });
});
