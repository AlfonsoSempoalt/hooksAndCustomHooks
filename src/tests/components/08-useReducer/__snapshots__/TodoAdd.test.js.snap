// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests to todoAdd should match with the snapshot 1`] = `
<Fragment>
  <h4>
    Add todo
  </h4>
  <hr />
  <form
    className="d-grid gap-2"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="learn..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-2 btn-block"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
